src/analysis/souffle/tests/test_taint_with_auth_logic.dl

```diff
@@ -0,0 +1,101 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#include "../taint_with_auth_logic.dl"
+#include "../flat_graph_ir.dl"
+#include "../edges_from_graph.dl"
+
+// This is a datalog file that is meant to test taint_with_auth_logic.dl. It 
+// should eventually be replaced by or otherwise used as a part of an
+// automated test.
+
+//-----------------------------------------------------------------------------
+// Program Manifests
+//-----------------------------------------------------------------------------
+
+// Module 1
+//      reads: 
+//          some_struct{ a, b }
+//          other_input
+//      writes:
+//          mod1out
+bindFlatParticle("module1", $FlatParticle(
+    /* reads */
+    $Cons("some_struct.a", $Cons("some_struct.b",
+        $Cons("other_input", $Nil()))),
+    /* writes */
+    $Cons("mod1out", $Nil())
+)).
+
+// the label of "other_input" is the empty set (which makes it public).
+// so it has no label elements
+
+// Module 2
+//     reads:
+//         mod1out
+//         input4
+//     writes:
+//         mod2out 
+bindFlatParticle("module2", $FlatParticle(
+    /* reads */
+    $Cons("mod1out", $Cons("input4", $Nil())),
+    /* writes */
+    $Cons("mod2out", $Nil())
+)).
+
+//-----------------------------------------------------------------------------
+// Policy
+//-----------------------------------------------------------------------------
+
+// Assign owners to tags
+ownsTag("Alice", "raw_video").
+ownsTag("Alice", "image_selection").
+ownsTag("Alice", "product_id").
+ownsTag("ServiceProvider", "model_weights").
+
+// Assign tags to initial inputs
+claimHasTag("some_struct.a", "raw_video").
+claimHasTag("some_struct.b", "image_selection").
+claimHasTag("input4", "model_weights").
+
+// The label of mod1out is downgraded to public (= empty set) 
+// because both the raw_video and image_selection tags are removed, and then
+// upgraded to {"product_id"} with this line:
+claimHasTag("mod1out", "product_id").
```

Comment:
"product_id" would be an integrity tag, right? How would we prevent this claim from being added by someone other than the owner?

Reply:
Re trusting these claims: see other comment thread with the header "ClaimHasTag has to be trusted"

No this is still an a confidentiality tag,  but I see how this is confusing. (There was a comment above this one to try to explain it, but maybe it did not help). Unlike the other "claimHasTag" tuples that set the labels of ingress inputs, this tuple is used to "reduce the scope of a downgrade". What we want to happen is for Alice to authorize the label of `mod1out` to be downgraded from `{raw_video, image_selection}` to just `{product_id}`. This happens in two steps:
   - 1) the label is downgraded from `{raw_video, image_selection}` to `{}` (where empty set is totally public -- it's allowed to go anywhere).
   - 2) the label is upgraded from `{}` to `{product_id}` because this is a label with more restrictions than totally public. (If/when we also add the "may" part of the policy language, the specific tag `product_id` might also permit different actions that the original tags)

Right now these two steps happen with two different relations, but I went back and forth with a few different language/api designs. It might be better to have the downgrade relation operate on labels, (with both a to and from label) but then implement this in terms of tag operations so that ownership of tags is checked -- (this means the individual requesting the downgrade needs to know the old label) so that the downgrade+upgrade combination can be done in one place. It seemed better to get to a working prototype faster and defer iteration like this until later.

Reply:
Just in case this still seems strange, here is slightly more explanation.

Downgrades weaken restrictions on data because they move the label lower in the lattice order (regardless of whether confidentiality or integrity is being downgraded), so they are controlled.

Upgrades add more restrictions to data because they move the label higher (for both conf/integ), so these can always safely be done.

With our label model based on (eventually pairs of sets):
For confidentiality -- downgrades remove tags, upgrades add tags
For integrity -- downgrades add tags, upgrades remove tags

So for integrity it would be fine to remove a tag, but adding one would need to be controlled.

Reply:
> So for integrity it would be fine to remove a tag, but adding one would need to be controlled.

Yes, this is what I meant. I really like the FLAM model of considering a tag to have a `secrecy` and `integrity` components. "product_id" can have both aspects. The integrity component would capture the fact that it is an untampered "product_id". The secrecy component would capture the fact that it just has the entropy of a product id and nothing more.